File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        root = None
        nodes = {}
        for p, c, isLeft in descriptions:
            if p in nodes:
                parent = nodes[p][0]
            else:
                parent = TreeNode(p)
                nodes[p] = (parent, False)
            if c in nodes:
                child = nodes[c][0]
                nodes[c] = (child, True)
            else:
                child = TreeNode(c)
                nodes[c] = (child, True)
            if isLeft:
                parent.left = child
            else:
                parent.right = child
        for node, hasParent in nodes.values():
            if not hasParent:
                root = node
                break
        return root

File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        TreeNode* root = nullptr;
        unordered_map<int, pair<TreeNode*, bool>> nodes;
        for(int i = 0; i < descriptions.size(); i++)
        {
            TreeNode* parent = nullptr;
            TreeNode* child = nullptr;
            if(nodes.count(descriptions[i][0]))
            {
                parent = nodes[descriptions[i][0]].first;
            }
            else
            {
                parent = new TreeNode(descriptions[i][0]);
                nodes[descriptions[i][0]] = {parent, false};
            }
            if(nodes.count(descriptions[i][1]))
            {
                child = nodes[descriptions[i][1]].first;
                nodes[descriptions[i][1]].second = true;
            }
            else
            {
                child = new TreeNode(descriptions[i][1]);
                nodes[descriptions[i][1]] = {child, true};
            }
            if(descriptions[i][2] == 1)
            {
                parent->left = child;
            }
            else
            {
                parent->right = child;
            }
        }
        for(auto& it : nodes)
        {
            if(it.second.second == false)
            {
                root = it.second.first;
                break;
            }
        }
        return root;
    }
};